src/paper.rs

```
@@ -3,45 +3,50 @@
 
 use crate::prelude::*;
 use chrono::prelude::*;
-use once_cell::sync::Lazy;
+#[cfg(test)]
+use mockito;
 use reqwest::header;
 use serde::Deserialize;
 use std::collections::HashMap;
-use std::env;
-use url::Url;
-
-pub static PAPER: Lazy<Paper> = Lazy::new(|| Paper { __priv: () });
 
-static API_DOMAIN: Lazy<String> = Lazy::new(|| {
-    env::var("PAPER_API_DOMAIN").unwrap_or_else(|_| String::from("https://papermc.io/api"))
-});
-static PROJECT_NAME: Lazy<String> =
-    Lazy::new(|| env::var("PAPER_PROJECT").unwrap_or_else(|_| String::from("paper")));
+use url::Url;
 
 const ACCEPT_JSON: &'static str = "application/json";
 
-pub struct Paper {
-    #[allow(dead_code)]
-    __priv: (),
+#[derive(Getters)]
+#[getset(get = "pub", set = "pub")]
+pub struct PaperApi {
+    domain: String,
+    project: String,
 }
 
-#[derive(Deserialize, Getters)]
+#[derive(Deserialize, Getters, Debug, PartialEq)]
 #[getset(get = "pub")]
 pub struct Project {
     version_groups: Vec<String>,
     versions: Vec<String>,
 }
 
-#[derive(Deserialize, Getters)]
+type BuildNumber = i32;
+#[derive(Deserialize, Getters, Debug, PartialEq)]
 #[getset(get = "pub")]
 pub struct Version {
-    builds: Vec<i32>,
+    builds: Vec<BuildNumber>,
+}
+
+impl Version {
+    pub fn get_latest_build_number(&self) -> BuildNumber {
+        *self
+            .builds
+            .last()
+            .expect("There should always be at least one build for version")
```

This is not an API guarantee.

---

I think we can reasonably expect it to exist. Though looking at bibliothek source code, it looks like it directly serves what's stored in mongo, which is updated by the CLI tool which has a race condition that would cause the whole request to fail, rather than get empty build list ( See: https://github.com/PaperMC/bibliothek/blob/main/cli/insertBuild.js#L90-L180 ).

I'll try to catch kashike to ask if such API guarantee can be added to bibliothek.

---

As discussed, not and API guarantee and even is possible due to race in the implementation of bibliothek CLI. Fixed.